### ISSUE TITLE
In progress detach operation is lost during pageserver restart [P:2] [S:3]

### ISSUE THREAD

**Issue by LizardWizzard** [neondatabase/neon#2238]
Can be solved with additional flags in timeline metadata.

The problem happens when pageserver answers with ok to detach or attach and then restarts before the operation succeeds

**Comment by hlinnaka** [comment]
see https://github.com/neondatabase/neon/pull/2239

**Comment by LizardWizzard** [comment]
Attach is no longer lost, still relevant for detach

**Comment by shanyp** [comment]
@ololobus do you retry these API calls ? what is the policy from the control plane on such failures ?

**Comment by ololobus** [comment]
> do you retry these API calls ? what is the policy from the control plane on such failures ?

- Detach tenant is used now when we delete project in the console. And actually we even ignore 500s now (see [1]), as previously pageserver was responding with 500 if tenant is absent locally. But this could a retry of tenant deletion operation, so we had no option to distinguish whether it's tenant not found or some actual internal error. BTW, is it still the case? If pageserver can respond with 404 now if there is no tenant locally, then we can change and retry all non 200 and 404 only
- Attach is only used in the migration code, and as far as I see it will retry everything non-202. See [2]. Existing code for migration is mostly broken and doesn't take into account many corner-cases like presence of many active computes. I have an issue for refactoring this control-plane code https://github.com/neondatabase/cloud/issues/3113. I think we can start working on it in parallel with planed Dmitry's changes on the pageserver, but it seems that we should sync with you and Dmitry first

[1] https://github.com/neondatabase/cloud/blob/5de42edae8adf572985268d82b39919b1360a012/goapp/internal/client/psclient/httppageserver/httppageserver.go#L259-L273
[2] https://github.com/neondatabase/cloud/blob/5de42edae8adf572985268d82b39919b1360a012/goapp/internal/client/psclient/httppageserver/httppageserver.go#L329-L333

**Comment by ololobus** [comment]
> 2. Attach wont be lost, this is already done.

How pageserver reacts on the repeated attach? Is it the same 202 or 500 or something else?

**Comment by LizardWizzard** [comment]
It will be 409 Conflict: Tenant X already exists

**Comment by ololobus** [comment]
@shanyp thanks! Created https://github.com/neondatabase/cloud/issues/4368 to do not forget

**Comment by LizardWizzard** [comment]
I agree with the analysis. It seems that the main issue was the absence of retried. 

However, I'm not sure that we can safely continue from interrupted detach operation. I e:

Console calls delete, pageserver starts to delete files, then crashes and restarts. Because file deletion order is not specified and there no special detach marker which can prevent pageserver from loading this tenant during startup we can have half broken tenant that wont be deleted by following retry from console.

We;ll have mark file for deletion of a tenant as described in the RFC. See https://github.com/zenithdb/zenith/blob/4158e24e60d294e0f039395ea95dd87f8ab317d9/docs/rfcs/022-pageserver-delete-from-s3.md#L76

I think we can close this issue, I can create another one in deletion project that mentions specifically out of order deletions. WDYT?

**Comment by problame** [comment]
> Console calls delete, pageserver starts to delete files, then crashes and restarts. Because file deletion order is not specified and there no special detach marker which can prevent pageserver from loading this tenant during startup we can have half broken tenant that wont be deleted by following retry from console.
> 
> We;ll have mark file for deletion of a tenant as described in the RFC. See https://github.com/zenithdb/zenith/blob/4158e24e60d294e0f039395ea95dd87f8ab317d9/docs/rfcs/022-pageserver-delete-from-s3.md#L76

Yeah, that's an issue but it's orthogonal to this one.

Closing this one. Please create a new issue and add it to your deletion project.
